sdk/python/lib/pulumi/next/output.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2016-2018, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import asyncio
+from inspect import isawaitable
+from typing import (
+    TypeVar,
+    Generic,
+    Set,
+    Callable,
+    Awaitable,
+    Union,
+    cast,
+    Mapping,
+    Any,
+    TYPE_CHECKING
+)
+
+from . import runtime
+if TYPE_CHECKING:
+    from .resource import Resource
+
+T = TypeVar('T')
+U = TypeVar('U')
+
+Input = Union[T, Awaitable[T], 'Output[T]']
+Inputs = Mapping[str, Input[Any]]
+
+
+class Output(Generic[T]):
+    """
+    Output helps encode the relationship between Resources in a Pulumi application. Specifically an
+    Output holds onto a piece of Data and the Resource it was generated from. An Output value can
+    then be provided when constructing new Resources, allowing that new Resource to know both the
+    value as well as the Resource the value came from.  This allows for a precise 'Resource
+    dependency graph' to be created, which properly tracks the relationship between resources.
+    """
+
+    _is_known: Awaitable[bool]
+    """
+    Whether or not this 'Output' should actually perform .apply calls.  During a preview,
+    an Output value may not be known (because it would have to actually be computed by doing an
+    'update').  In that case, we don't want to perform any .apply calls as the callbacks
+    may not expect an undefined value.  So, instead, we just transition to another Output
+    value that itself knows it should not perform .apply calls.
+    """
+
+    _future: Awaitable[T]
+    """
+    Future that actually produces the concrete value of this output.
+    """
+
+    _resources: Set[Resource]
+    """
+    The list of resources that this output value depends on.
+    """
+
+    def __init__(self, resources: Set[Resource], future: Awaitable[T], is_known: Awaitable[bool]) -> None:
+        self._resources = resources
+        self._future = future
+        self._is_known = is_known
+
+    def resources(self) -> Set[Resource]:
+        return self._resources
+
+    def future(self) -> Awaitable[T]:
+        return self._future
```

Comment:
this is effectively supposed to be private in TS.  is that the intent in python as well?